client/client.go

```diff
@@ -630,6 +685,66 @@ func newDefaultRequester(client *Client, opts *Config) (*defaultRequester, error
 	return requester, nil
 }
 
+// verifyConnection verifies the incoming server TLS certificate.
+func verifyConnection(state tls.ConnectionState, opts *Config) error {
+	// We always expect two certificates from our server:
+	//
+	// state.PeerCertificates[0] - server TLS certificate
+	// state.PeerCertificates[1] - server Identity certificate (root CA)
+	certCount := len(state.PeerCertificates)
+	if certCount != expectedServerCertCount {
+		return fmt.Errorf("cannot find identity certificate: expected %d certificates, got %d", expectedServerCertCount, certCount)
+	}
+	// Make a local copy of the server identity certificate (root CA).
+	serverIDCert := state.PeerCertificates[1]
+
+	// Only one TLS server option is allowed.
+	tlsOptions := 0
+	if opts.TLSServerFingerprint != "" {
+		tlsOptions += 1
+	}
+	if opts.TLSServerIDCert != nil {
+		tlsOptions += 1
+	}
+	if opts.TLSServerInsecure {
+		tlsOptions += 1
+	}
+	if tlsOptions > 1 {
+		return fmt.Errorf("only one TLS server validation option allowed, but %d were provided", tlsOptions)
+	}
+
+	if opts.TLSServerFingerprint != "" {
+		// Client supplied fingerprint must match the server identity.
+		idFingerprint, err := getIdentityFingerprint(serverIDCert)
+		if err != nil {
+			return fmt.Errorf("cannot obtain identity fingerprint: %v", err)
+		}
+		if idFingerprint == opts.TLSServerFingerprint {
```

Comment:
Should we use https://pkg.go.dev/crypto/subtle#ConstantTimeCompare here?

Reply:
We also omitted this in the past TLS PRs for similar reasons - will confirm with Ben.

Reply:
Discussed, resolved.